### ISSUE TITLE
[STORM-1971] HDFS Timed Synchronous Policy

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5753]
<p>When the data need to be wrote to HDFS is not very large in quantity . We need a timed synchronous policy to flush cached date into HDFS periodically.</p>

---
<details><summary><i>Originally reported by <a title="darion yaphet" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=darion">darion</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1971" target="_blank">HDFS Timed Synchronous Policy</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="darion yaphet" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=darion">darion</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2024-07-02T10:47:08+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user darionyaphet opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1566" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1566</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1971" title="HDFS Timed Synchronous Policy" class="issue-link" data-issue-key="STORM-1971"><del>STORM-1971</del></a> HDFS Timed Synchronous Policy</p>

<p>    <span class="error">&#91;STORM-1971 HDFS Timed Synchronous Policy&#93;</span>(<a href="https://issues.apache.org/jira/browse/STORM-1971" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-1971</a>)</p>

<p>    When the data need to be wrote to HDFS is not very large in quantity . </p>

<p>    We need a timed synchronous policy to flush cached date into HDFS periodically.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/darionyaphet/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/darionyaphet/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-1971" title="HDFS Timed Synchronous Policy" class="issue-link" data-issue-key="STORM-1971"><del>STORM-1971</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1566.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1566.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1566</p>

<hr />
<p>commit 86e036a8544040557f6b514e0d27f119c6473727<br/>
Author: darionyaphet <darion.yaphet@gmail.com><br/>
Date:   2016-07-14T13:34:45Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1971" title="HDFS Timed Synchronous Policy" class="issue-link" data-issue-key="STORM-1971"><del>STORM-1971</del></a> : HDFS Timed Synchronous Policy</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=darion">darion</a>:</i>
<p>Hi , I checked default.yaml and found topology.tick.tuple.freq.secs is null . So if I don't specify this item in config the tick tuple will not send to bolts .</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett">dossett</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=darion" class="user-hover" rel="darion">darion yaphet</a> There is a default value for the tick tuples, 15 seconds.<br/>
<a href="https://github.com/apache/storm/blob/master/external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/AbstractHdfsBolt.java#L57" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/AbstractHdfsBolt.java#L57</a></p>

<p>The tick tuple approach guarantees a sync at a defined interval.  A solution based on SyncPolicy can't do that because it can only trigger when a tuple is being processed.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1566#discussion_r71151878" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1566#discussion_r71151878</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/sync/CountSyncPolicy.java &#8212;<br/>
    @@ -26,7 +26,7 @@</p>
<ul>
	<li>have been processed.<br/>
      */<br/>
     public class CountSyncPolicy implements SyncPolicy {</li>
</ul>
<ul class="alternate" type="square">
	<li>private int count;<br/>
    +    private final int count;
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    Great catch to make that final. I would be +1 on that change in a separate PR.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user darionyaphet commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1566#discussion_r71182808" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1566#discussion_r71182808</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/sync/CountSyncPolicy.java &#8212;<br/>
    @@ -26,7 +26,7 @@</p>
<ul>
	<li>have been processed.<br/>
      */<br/>
     public class CountSyncPolicy implements SyncPolicy {</li>
</ul>
<ul class="alternate" type="square">
	<li>private int count;<br/>
    +    private final int count;
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    @dossett OK fixed <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jakesjohn">jakesjohn</a>:</i>
<p> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett" class="user-hover" rel="dossett">Aaron Blake Niskode-Dossett</a>  When I read the code, I find that each tuple is written to HDFS one by one when it is processed even though filesystem sync is called at Tick_tuple_frequency. Shouldn't the write happen only for the tuplebatch at periodic intervals(may be sync interval) for minimizing the network cost and for higher throughput? Why is each tuple written one by one to hdfs? <br/>
<a href="https://github.com/apache/storm/blob/f48d7941b10483e87a30b4849321c4dc0844a5a5/external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/AbstractHdfsBolt.java#L152" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/f48d7941b10483e87a30b4849321c4dc0844a5a5/external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/AbstractHdfsBolt.java#L152</a> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett">dossett</a>:</i>
<p>Writing each tuple doesn't necessarily result in higher network costs since caching can happen in HDFS code.  (I am not an HDFS expert, feel free to correct me if I'm wrong).  I view the sync as an outer limit on when the data is guaranteed to have been flushed and persisted to HDFS.</p>

<p>It is possible that an alternative approach (batching the writes) would result in better performance &#8211; I'd be interested in seeing benchmarks for that.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jakesjohn">jakesjohn</a>:</i>
<p>Thanks for your reply.    I think overhead is very high in the current case. Writing a message of x size y times to a hdfs cluster vs write a single message of xy size to hdfs cluster.  HDFS is best at large streaming reads and writes  What are the advantages of current implementation(one message write) over the proposed one(batching writes)?    </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett">dossett</a>:</i>
<p>I did not create that approach (although I did carry it over in a re-write) but one advantage that comes to mind is that data will potentially appear in HDFS sooner.  Also, not all implementations of AbstractHdfsBolt may be able to support the "write one big message" approach.  A text file bolt certainly could, but what about Sequence File or Avro?  I don't know off the top of my head.</p>

<p>If you pursue that and are able to contribute a patch, I will definitely be available for review.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jakesjohn">jakesjohn</a>:</i>
<p>Can we say that data will be in HDFS sooner?  Readers can see the latest message writes only after sync. Hence, i feel that it is safe to write a batch and do a sync immediately during every tick.   As you said, this will work for text files and i need to investigate Avro and Sequence. I am ready to contribute and I will take this up very soon.   Thanks</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rzo1">rzo1</a>:</i>
<p>Storm 1.x isn't maintained anymore. If the issue still applies for 2.6.x, feel free to re-open.</p>
